### ISSUE TITLE
Auto Gather/Build #2 can be set to build without actually buying a trap

### ISSUE THREAD

**Issue by ElderThorin** [genbtc/AutoTrimps#36]
Auto Gather/Build #2 gets stuck selecting 'build' without actually buying any traps to build, so it just has build pointlessly selected.

This is for very early zones prior to auto traps being found.

**Comment by genbtc** [comment]
This is likely a case of your low level earlygame priorities not matching what needs to happen at normal/higher levels. In that case, use #1 which I assume works. Either I am misunderstanding, or you are. "Pointlessly Selected" sounds like a misnomer, because if more than 2 buildings are in the build queue, it should be on "Build". Anyone at normal levels of the game cannot keep up with passively Slow-building things.  There is no way to determine whether you will agree with this or not, thats why there is an independent choice of #2 at all. What was wrong with #1 that you went to #2 ?
On another tangent, there is a line in the code that asks if you have "AutoTraps on" and less than 1000 traps, then also actively Build, which would be the thing I expect you to mention next. I may consider decreasing the seemingly arbitrary number, but I dont believe theres anything I can do about what you mention, unless you have an idea.

**Comment by genbtc** [comment]
Rereading, "gets stuck without actually buying any traps" - how can this happen, The conditions to buy traps are: "TrapTrimps" setting is on, (breedingTrimps < 5 OR time remaining it takes to breed is greater than "GeneticistTimer" setting), Traps owned equals 0, and can afford 1 trap.  If all those are true, it will queue a trap, not exit the function, then ask the build-queue questions, so it can build it. So i dont know where that code went wrong.

**Comment by ElderThorin** [comment]
The buy traps and build traps conditions are a little different, and it ends up sitting there 'building' with nothing in the build queue.  In the Auto Gather/Build 1 method it puts the two together, presumably to avoid this type of issue?  If you read over the logic you can see where it's possible to end up at setting 'building' for traps without actually checking that anything is in the build queue.

The third condition on doing setGather('buildings') will set it to build if there are < 1000 traps even if the earlier check to buy traps ended up not buying them, so if you have less than 1000 traps, it will just sit there trying to build traps that haven't been bought.

**Comment by genbtc** [comment]
I did read over the code and I dont see it. Show me how it can build with something not in the build queue ? 
OH. are you talking about building as in TRAPPING ? Building the build-queue is setGather('buildings')  - this is the Huts/houses/traps.  setGather('trimps') is TRAPPING. (which is only set to happen if you have >0 traps), when you are down to 0, it builds more. SO i still have no idea what you mean by stuck.

**Comment by ElderThorin** [comment]
Sorry if I haven't been very clear, I'm mostly busy with work ;-).  Let me try again.

The problem I observed is Auto Build/Gather 2 is setting 'building' when nothing is in the build queue.  This was happening early in a run where foremen do not automatically buy+build traps.

I did a quick scan of gather.js manual labor 2 method and it seemed most likely this is due to the following bits of code, but I only looked briefly and I could be incorrect on the reason:

```
    //Traps and Trimps
    if (trapTrimpsOK && (breedingTrimps < 5 || targetBreed < getBreedTime(true))) {
        if (game.buildings.Trap.owned > 0) {
            setGather('trimps');//gatherTrimps = true;
            return;
        }
        if (game.buildings.Trap.owned == 0 && canAffordBuilding('Trap'))
            safeBuyBuilding('Trap');//buyTraps = true;
    }
    //Buildings:
    //if we have more than 2 buildings in queue, or (our modifier is real fast and trapstorm is off), build
    if ((!game.talents.foreman.purchased && (game.global.buildingsQueue.length ? (game.global.buildingsQueue.length > 1 || game.global.autoCraftModifier == 0 || (getPlayerModifier() > 1000 && game.global.buildingsQueue[0] != 'Trap.1')) : false)) ||
    //if trapstorm is off (likely we havent gotten it yet, the game is still early, buildings take a while to build ), then Prioritize Storage buildings when they hit the front of the queue (should really be happening anyway since the queue should be >2(fits the clause above this), but in case they are the only object in the queue.)
    (!game.global.trapBuildToggled && (game.global.buildingsQueue[0] == 'Barn.1' || game.global.buildingsQueue[0] == 'Shed.1' || game.global.buildingsQueue[0] == 'Forge.1')) ||
    //Build more traps if we have TrapTrimps on, and we own less than 1000 traps.
    (trapTrimpsOK && game.global.trapBuildToggled && game.buildings.Trap.owned < 1000)) {
        setGather('buildings');//buildBuildings = true;
        return;
    }
```

You can see that the first if statement calls safeBuyBuilding('Trap') if none are owned, it can be afforded, and some other bits I didn't dig into. 

In the second if, the condition (trapTrimpsOK && game.global.trapBuildToggled && game.buildings.Trap.owned < 1000) being true causes setGather('buildings').  I believe this is the reason for the behavior I observed (ie, it looks like it is possible for this to setGather('buildings') even if the earlier if didn't actually result in buying any traps (and that condition looks designed to set gather for buildings specifically for traps since it doesn't check the build queue length or what is in the build queue.)

**Comment by genbtc** [comment]
See im a mindreader, I knew this would come up, but you said "This is for very early zones prior to auto traps being found".  game.global.trapBuildToggled is the variable for the AutoTraps button being on. So how can it be ? 

I'll re-analyze everything under the presumption that you did have Autotraps on.

**Comment by ElderThorin** [comment]
That I can't answer.  Is it maybe hidden but still on from the previous run?  Haven't looked closely enough. Or maybe I'm completely wrong about the reason :-)

**Comment by genbtc** [comment]
Wow you're right. It is hidden and still on from the previous portal. Good catch

**Comment by genbtc** [comment]
Ive re-done AutoGather&Build 2's build code entirely ab3a03b667730a42bb41efe429fabca4b7c5ba7d.
```
AutoGather #2 changes to "build": 
//Buildings:
    //pre-requisites for all: have something in the build queue, and playerCraftmod does actually speed it up..
//if we have 2 or more buildings in queue, and playerCraftmod is high enough (>3x autoCraftmod) to speed it up.
//Prioritize Storage buildings when they hit the front of the queue (in case they are the only object in the queue).
//manualBuild traps if we have TrapTrimps on, AutoTraps on (and available), and we own less than 100 traps.
```
EDIT: needed some boolean parentheses (nm, It didnt actually need it -just being careful)

**Comment by genbtc** [comment]
Thank you for your help.
